### PR TITLE
4897 dataset commands

### PR DESCRIPTION
**Pull request by mannSheth**
The code seems to function as described in discussions.

***
### Would love feedback in regards to method names and basic messages that are thrown if change is needed.
***

_This is blocked as it needs the data layer change to function._

### Async:
image

### Blocking:
image

### CODE REVIEW

**Comment by mannSheth** [comment]
most probably a good idea to just review but not merge as it kinda of does nothing without the rest of the dataset code

File: austrakka/components/project/dataset/__init__.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    ctx.context = ctx.parent.context
+
+
+@dataset.command('async-add')
+@click.option('-fp',
+              '--file-path',
+              help='dataset csv file to upload to the project', )
+@opt_analysis_label()
+@opt_abbrev(help="Project Abbreviation")
+def dataset_add(
```
Comment (viatunimelb) [inline austrakka/components/project/dataset/__init__.py:22]: Should include @table_format_option() like other commands do.

File: austrakka/components/project/dataset/funcs.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        )  # Replace this with your actual function to fetch status
+
+        if status_change is not None:
+            logger.success(f"Current status: {status_change}")
+
+            if status_change == 'Finished':
+                logger.success('')
+                break  # Exit the loop when the desired status is reached
+            if "Failed" in status_change:
+                logger.error("The Job in-jest has failed")
```
Comment (viatunimelb) [inline austrakka/components/project/dataset/funcs.py:123]: Typo. Should be "ingest" not "in-jest".

File: austrakka/components/project/dataset/__init__.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        file_path: str,
+        analysis_label: str,
+        abbrev: str):
+    """Upload a dataset file to the given project in AusTrakka."""
+    add_dataset(file_path, analysis_label, abbrev)
+
+
+@dataset.command('track-job')
+@opt_abbrev(help="Project Abbreviation")
```
Comment (claresloggett) [inline austrakka/components/project/dataset/__init__.py:31]: I'm wary of calling these "jobs" in a user-facing way (i.e. the command name track-job). I think for binfies, jobs are intuitively linked to data analysis, which is also a meaning we're later going to want to assign them. Could we call this "track-dataset"?

File: austrakka/components/project/dataset/funcs.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    path = "/".join([PROJECT_PATH,
+                     abbrev,
+                     DATASET_TRACK_DETAILED_PATH if detailed else DATASET_TRACK_PATH,
+                     tracking_token])
+    if detailed:
+        response = api_get(path)
+        data = response['data'] if ('data' in response) else response
+        if not data:
+            logger.info("No JobFeedbacks available")
+            return
```
Comment (claresloggett) [inline austrakka/components/project/dataset/funcs.py:63]: Under what circumstances will there be no JobFeedbacks? No recent uploads? No unfinished data ingestions?

I'm thinking of avoiding "job" in the message, but I'm not sure what the message really signifies in practice.

File: austrakka/components/project/dataset/funcs.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        if status_change is not None:
+            logger.success(f"Current status: {status_change}")
+
+            if status_change == 'Finished':
+                logger.success('')
+                break  # Exit the loop when the desired status is reached
+            if "Failed" in status_change:
+                logger.error("The Job in-jest has failed")
+                break
```
Comment (claresloggett) [inline austrakka/components/project/dataset/funcs.py:124]: Can we change error to "The dataset ingest has failed"? (Notice also typo in ingest)

File: austrakka/utils/options.py [hunk 6]
```diff
@@ -77,6 +77,31 @@ def opt_abbrev(**attrs: t.Any):
     )
 
 
+def opt_tracking_token(**attrs: t.Any):
+    defaults = {
+        'required': True,
+        'help': 'tracking token of a job, that can be used to query its status'
+    }
```
Comment (claresloggett) [inline austrakka/utils/options.py:84]: Avoiding "job" OR be very specific - "tracking token for a dataset upload job, ..."

**Review by claresloggett — CHANGES_REQUESTED** [review]

File: austrakka/components/project/setup/funcs.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    # Replace 'abbrev' with your actual value
+    return api_patch(
+        path=f'{PROJECT_PATH}/add-project-field/{abbrev}',
+        data=field_name_objs
+    )
+
+
+@logger_wraps()
+def set_merge_algorithm_project(abbrev: str, merge_algorithm: str):
+    if merge_algorithm == 'show-all':
```
Comment (samcarswell) [inline austrakka/components/project/setup/funcs.py:29]: Why not set `opt_merge_algorithm` `click.Choice` options to `ShowAll` and `Override` so we don't need to change what's passed in

File: austrakka/components/project/dataset/__init__.py (hunk 3 above)
Comment (viatunimelb) [reply austrakka/components/project/dataset/__init__.py:36]: I am changing it to "track-upload". The full command would be "austrakka project dataset track-upload"

File: austrakka/components/project/setup/funcs.py (hunk 7 above)
Comment (viatunimelb) [reply austrakka/components/project/setup/funcs.py:29]: I think the lower case and dash notation suits the current language (other command sets) more than capital-case. In the interest of merging this with the least changes possible, lets go with this for now. We can always change it later.

**Review by viatunimelb — APPROVED** [review]

File: austrakka/components/project/dataset/funcs.py (hunk 4 above)
Comment (viatunimelb) [reply austrakka/components/project/dataset/funcs.py:67]: If the dataset as been uploaded but not acknowledged (using add-async) command. Feedback starts only from the validation stage onward.

File: austrakka/components/project/setup/__init__.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    opt_field_and_source, \
+    opt_abbrev, opt_prov_id
+from austrakka.utils.output import table_format_option
+
+
+# pylint: disable=duplicate-code
+@click.group()
+@click.pass_context
+def setup(ctx):
+    """Commands to upload project datasets"""
```
Comment (claresloggett) [inline austrakka/components/project/setup/__init__.py:21]: This is an error - the command group `setup` has the same text as `dataset`.

However I actually think (as per slack discussion) we should split these commands into two command groups: `provision` and `field`, and do away with `setup` altogether

**Review by claresloggett — CHANGES_REQUESTED** [review]

**Review by claresloggett — APPROVED** [review]

**Pull request merged**